File: client/src/main/scala/stasis/client/ops/backup/stages/EntityCollection.scala
package stasis.client.ops.backup.stages

import org.apache.pekko.NotUsed
import org.apache.pekko.stream.scaladsl.Flow

import stasis.client.collection.BackupCollector
import stasis.client.model.SourceEntity
import stasis.client.ops.Metrics
import stasis.client.ops.backup.Providers
import stasis.shared.model.datasets.DatasetDefinition
import stasis.shared.ops.Operation

trait EntityCollection {
  protected def targetDataset: DatasetDefinition
  protected def providers: Providers

  private val metrics = providers.telemetry.metrics[Metrics.BackupOperation]

  def entityCollection(implicit operation: Operation.Id): Flow[BackupCollector, SourceEntity, NotUsed] =
    Flow[BackupCollector]
      .flatMapConcat(_.collect())
      .wireTap { entity =>
        metrics.recordEntityExamined(entity = entity)
        providers.track.entityExamined(
          entity = entity.path,
          metadataChanged = entity.hasChanged,
          contentChanged = entity.hasContentChanged
        )
      }
      .wireTap { entity =>
        if (entity.hasChanged) {
          metrics.recordEntityCollected(entity = entity)
          providers.track.entityCollected(entity = entity)
        } else {
          metrics.recordEntitySkipped(entity = entity)
          providers.track.entitySkipped(entity = entity.path)
        }
      }
      .filter(_.hasChanged)
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/backup/stages/EntityCollection.kt
package stasis.client_android.lib.ops.backup.stages

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.flatMapConcat
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.onEach
import stasis.client_android.lib.collection.BackupCollector
import stasis.client_android.lib.model.SourceEntity
import stasis.client_android.lib.model.server.datasets.DatasetDefinition
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.ops.backup.Providers

interface EntityCollection {
    val targetDataset: DatasetDefinition
    val providers: Providers

    @OptIn(ExperimentalCoroutinesApi::class)
    fun entityCollection(
        operation: OperationId,
        flow: Flow<BackupCollector>
    ): Flow<SourceEntity> =
        flow
            .flatMapConcat { collector -> collector.collect() }
            .onEach { entity ->
                providers.track.entityExamined(
                    operation = operation,
                    entity = entity.path
                )
            }
            .map { entity ->
                if (entity.hasChanged) {
                    providers.track.entityCollected(operation = operation, entity = entity)
                } else {
                    providers.track.entitySkipped(operation = operation, entity = entity.path)
                }

                entity
            }
            .filter { it.hasChanged }
}
